background/content-scripts.js

```diff
@@ -54,29 +55,51 @@ const contentScripts = (() => {
 
   function injectToAllTabs() {
     return queryTabs({}).then(tabs => {
-      const busyTabs = new Set();
       for (const tab of tabs) {
-        // skip unloaded/discarded tabs
-        if (!tab.width || tab.discarded) continue;
+        // skip unloaded/discarded/chrome tabs
+        if (!tab.width || tab.discarded || !URLS.supported(tab.url)) continue;
         // our content scripts may still be pending injection at browser start so it's too early to ping them
         if (tab.status === 'loading') {
-          busyTabs.add(tab.id);
+          trackBusyTab(tab.id, true);
         } else {
           injectToTab({
             url: tab.url,
             tabId: tab.id
           });
         }
       }
-      if (busyTabs.size) {
-        chrome.tabs.onUpdated.addListener(function _(tabId, {status}, {url}) {
-          if (status === 'complete' && busyTabs.has(tabId)) {
-            busyTabs.delete(tabId);
-            if (!busyTabs.size) chrome.tabs.onUpdated.removeListener(_);
-            injectToTab({tabId, url});
-          }
-        });
-      }
     });
   }
+
+  function toggleBusyTabListeners(state) {
+    const toggle = state ? 'addListener' : 'removeListener';
+    chrome.webNavigation.onCompleted[toggle](onBusyTabUpdated);
+    chrome.webNavigation.onErrorOccurred[toggle](onBusyTabUpdated);
+    chrome.webNavigation.onTabReplaced[toggle](onBusyTabUpdated);
+    chrome.tabs.onRemoved[toggle](onBusyTabRemoved);
+    if (state) {
+      busyTabs.timer = setTimeout(toggleBusyTabListeners, 15e3, false);
```

Comment:
I think we can use a separate variable for this.

Reply:
I hoped you wouldn't notice this :-)